vortexor/src/stake_updater.rs

```diff
@@ -0,0 +1,97 @@
+    std::{
+        collections::HashMap,
+        str::FromStr,
+        sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc, RwLock,
+        },
+        thread::{self, sleep, Builder, JoinHandle},
+        time::{Duration, Instant},
+    },
+};
+
+// The interval to refresh the stake information.
+const STAKE_REFRESH_INTERVAL: Duration = Duration::from_secs(5);
+
+/// This service is responsible for periodically refresh the stake information
+/// from the network with the assistance of the RpcLoaderBalancer.
+pub struct StakeUpdater {
+    thread_hdl: JoinHandle<()>,
+}
+
+impl StakeUpdater {
+    pub fn new(
+        exit: Arc<AtomicBool>,
+        rpc_load_balancer: Arc<RpcLoadBalancer>,
+        shared_staked_nodes: Arc<RwLock<StakedNodes>>,
+        staked_nodes_overrides: Arc<HashMap<Pubkey, u64>>,
+    ) -> Self {
+        let thread_hdl = Builder::new()
+            .name("stkUpdtr".to_string())
+            .spawn(move || {
+                let mut last_stakes = Instant::now();
+                while !exit.load(Ordering::Relaxed) {
+                    if let Err(err) = Self::try_refresh_stake_info(
+                        &mut last_stakes,
+                        &shared_staked_nodes,
+                        staked_nodes_overrides.clone(),
+                        &rpc_load_balancer,
+                    ) {
+                        warn!("Failed to refresh pubkey to stake map! Error: {:?}", err);
+                        sleep(STAKE_REFRESH_INTERVAL);
+                    }
+                }
+            })
+            .unwrap();
+
+        Self { thread_hdl }
+    }
+
+    /// Update the stake info when it has elapsed more than the
+    /// STAKE_REFRESH_INTERVAL since the last time it was refreshed.
+    fn try_refresh_stake_info(
+        last_refresh: &mut Instant,
+        shared_staked_nodes: &Arc<RwLock<StakedNodes>>,
+        staked_nodes_overrides: Arc<HashMap<Pubkey, u64>>,
+        rpc_load_balancer: &Arc<RpcLoadBalancer>,
+    ) -> client_error::Result<()> {
+        if last_refresh.elapsed() > STAKE_REFRESH_INTERVAL {
+            let client = rpc_load_balancer.rpc_client();
+            let vote_accounts = client.get_vote_accounts()?;
+
+            let stake_map = Arc::new(
+                vote_accounts
+                    .current
+                    .iter()
+                    .chain(vote_accounts.delinquent.iter())
+                    .filter_map(|vote_account| {
+                        Some((
+                            Pubkey::from_str(&vote_account.node_pubkey).ok()?,
+                            vote_account.activated_stake,
+                        ))
+                    })
+                    .collect::<HashMap<Pubkey, u64>>(),
+            );
+
+            *last_refresh = Instant::now();
+            let shared: StakedNodes = StakedNodes::new(stake_map, staked_nodes_overrides.clone());
+            *shared_staked_nodes.write().unwrap() = shared;
```

Comment:
We have a `RwLock` on the `StakedNodes`. We should just modify the stake map, not overwrite the whole object. Clear then extend from this iterator - it will re-use the allocation. No need to wrap `staked_nodes_overrides` in an Arc or clone it anywhere.

Reply:
The StakedNodes has an Arc to the stake_map HashMap. Even when we have the RwLock on StakedNodes, we cannot just modify the stake map in it

Reply:
It will be non-trivial to support stakes map in StakedNodes as it is tied to EpochStakes in the validator's side's code. I am planning to update the refresh interval for stakes as it is not frequently updated in any case on the validator -- per epoch. So the additional overhead won't be that impactful.